arrows/core/parameterize_mesh.cxx

```diff
@@ -0,0 +1,294 @@
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *  * Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ *
+ *  * Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ *  * Neither name of Kitware, Inc. nor the names of any contributors may be used
+ *    to endorse or promote products derived from this software without specific
+ *    prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * \file
+ * \brief Implementation of the mesh uv parameterization
+ */
+
+#include "parameterize_mesh.h"
+
+#include <algorithm>
+#include <iostream>
+#include <numeric>
+#include <Eigen/Dense>
+#include <vital/exceptions.h>
+
+using namespace kwiver::vital;
+
+namespace {
+
+/// This structure is used to represent a 2D triangle
+struct triangle_t
+{
+  kwiver::vital::vector_2d a;
+  kwiver::vital::vector_2d b;
+  kwiver::vital::vector_2d c;
+  unsigned int face_id;
+  unsigned int height;
+  unsigned int width;
+};
+}
+
+namespace kwiver {
+namespace arrows {
+namespace core {
+
+/// Private implementation class
+class parameterize_mesh::priv
+{
+public:
+  /// Constructor
+  priv()
+  {
+  }
+
+  priv(const priv& other)
+  {
+  }
+
+  double resolution;
+  unsigned int interior_margin;
+  unsigned int exterior_margin;
+};
+
+// Constructor
+parameterize_mesh::parameterize_mesh()
+  : d_(new priv)
+{
+  attach_logger( "arrows.core.parameterize_mesh" );
+}
+
+// Copy constructor
+parameterize_mesh
+::parameterize_mesh(const parameterize_mesh& other)
+  : d_(new priv(*other.d_))
+{
+}
+
+
+// Destructor
+parameterize_mesh
+::~parameterize_mesh()
+{
+}
+
+
+// Get this algorithm's \link vital::config_block configuration block \endlink
+vital::config_block_sptr
+parameterize_mesh::get_configuration() const
+{
+  vital::config_block_sptr config =
+      vital::algo::parameterize_mesh::get_configuration();
+  config->set_value("resolution", d_->resolution,
```

Comment:
I don't think that resolution really matters here.  The UV coordinates are still between 0 and 1 and should be unaffected by this parameter.  This parameter only scales the width and height return values.  I suggest removing those return values anyway (see other related comment).

Reply:
I am using the resolution to have the triangles in pixel units and to be able to set margins (in pixels) between them.